File: 3622-check-divisibility-by-digit-sum-and-product/3622-check-divisibility-by-digit-sum-and-product.cpp
class Solution {
private:
    pair<int, int> dig(int num) {
        if (num == 0)
            return {0, 1};

        int r = num % 10;

        auto p = dig(num / 10);

        return {p.first + r, p.second * r};
    }

public:
    bool checkDivisibility(int n) {
        return n % (dig(n).first + dig(n).second) == 0;
    }
};

File: 3622-check-divisibility-by-digit-sum-and-product/3622-check-divisibility-by-digit-sum-and-product.java
class Solution {
    private int digitSum(int num) {
        if (num == 0)
            return 0;
        return (num % 10) + digitSum(num / 10);
    }

    private int digitProduct(int num) {
        if (num == 0)
            return 1;

        return (num % 10) * digitProduct(num / 10);
    }

    public boolean checkDivisibility(int n) {
        return (n % (digitSum(n) + digitProduct(n)) == 0);
    }
}